backends/eopkg/eopkgBackend.py

```diff
@@ -113,17 +113,61 @@ def _load_settings(self):
         else:
             self.groups = {}
 
+    def progress_cb(self, **kw):
+        # we continue to get callbacks at 100%
+        # hijacking the progress bar so cap to 99.
+        if int(kw['percent']) < 99:
+            self.percentage(int(kw['percent']))
+
+    def status_cb(self, event, **keywords):
+        # FIXME: when getting repo, we can't use self.packagedb here as it interferes with atomicoperations
+        if event == pisi.ui.downloading:
+            self.status(STATUS_DOWNLOAD)
+            pkg = keywords["package"]
+            if self.packagedb.has_package(pkg.name):
+                repo = pisi.db.packagedb.PackageDB().which_repo(pkg.name)
+            else:
+                repo = "local"
+            pkg_id = self.get_package_id(pkg.name, pkg.version, pkg.architecture, repo)
+            self.package(pkg_id, INFO_DOWNLOADING, pkg.summary)
+        if event == pisi.ui.installing:
+            self.status(STATUS_INSTALL)
+            pkg = keywords["package"]
+            if self.packagedb.has_package(pkg.name):
+                repo = pisi.db.packagedb.PackageDB().which_repo(pkg.name)
+            else:
+                repo = "local"
+            pkg_id = self.get_package_id(pkg.name, pkg.version, pkg.architecture, repo)
+            self.package(pkg_id, INFO_INSTALLING, pkg.summary)
+        if event == pisi.ui.removing:
+            self.status(STATUS_REMOVE)
+            pkg = keywords["package"]
+            if self.packagedb.has_package(pkg.name):
+                repo = pisi.db.packagedb.PackageDB().which_repo(pkg.name)
+            else:
+                repo = "installed"
+            pkg_id = self.get_package_id(pkg.name, pkg.version, pkg.architecture, repo)
+            self.package(pkg_id, INFO_REMOVING, pkg.summary)
```

Comment:
Maybe I'm not seeing something obvious, but this looks like a bunch of duplicated code. Can we do only the event-specific operations in the if/else, and the common operations outside of the if/else?

Reply:
I thought the same but the event needs to be non null and the repo status can be slightly different for local pkgs for different events which in turn leads up up to it not really saving that many loc. If I do find a nice way I'll implement it.